compiler/rustc_session/src/config.rs

```diff
@@ -2451,6 +2451,19 @@ pub fn parse_externs(
             Some((opts, name)) => (Some(opts), name.to_string()),
         };
 
+        if !crate::utils::is_ascii_ident(&name) {
+            let mut error = handler.early_struct_error(format!(
+                "crate name `{name}` passed to `--extern` is not a valid ASCII identifier"
+            ));
+            let adjusted_name = name.replace("-", "_");
+            if crate::utils::is_ascii_ident(&adjusted_name) {
+                error.help(format!(
+                    "consider replacing the dashes with underscores: `{adjusted_name}`"
+                ));
+            }
+            error.emit();
+        }
```

Comment:
The lints `uncommon_codepoints`, `mixed_script_confusables` and `confusable_idents` are amazing for sure! I'd lean towards not doing that here since it'd mean moving the code from `rustc_lint` here and adapting it to an environment that doesn't have a `Session` yet, yuck! Not sure how easy that'd be, maybe it's not that bad. `early_errors` can't even be translated right now.

Reply:
@fmease fair points :+1: . It can also be later work, too.

Reply:
My stance is always "don't block on improving a solution short of perfect as long as it improves over the status quo", and this PR comfortably glides over that bar.